### PR TITLE
FIR/IIR Filters and RAVE prefilter example

### PR DESCRIPTION
**Pull request by hyakuchiki**
FIR/IIR filters for prefiltering input audio to models.
Performance should be fine for both, although FIR seems to be faster.
Didn't get around to FFTFIR (faster than standard FIR) because of some caching issues.

### CODE REVIEW

**Comment by hyakuchiki** [comment]
actually, group delay of the fir might need to be reported? (around 128 samples by default)

**Comment by hyakuchiki** [comment]
No, this delay is from the group delay of the filter which is causal and linear phase.
https://www.allaboutcircuits.com/technical-articles/finite-impulse-response-filter-design-by-windowing-part-i-concepts-and-rect/
so this delay should be filt_size // 2
not quite sure what happens when filt_size is even tho...

File: neutone_sdk/filters.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+
+        # HP = x - 2R*BP - LP
+        y_hps = audio - twoR.squeeze(-1) * Y[:, :, 0] - Y[:, :, 1]
+
+        y_mixed = (
+            twoR.squeeze(-1) * mix[:, :, 0] * Y[:, :, 0]
+            + mix[:, :, 1] * Y[:, :, 1]
+            + mix[:, :, 2] * y_hps
+        )
+        y_mixed = y_mixed.permute(1, 0).contiguous()
```
Comment (bogdanteleaga) [inline neutone_sdk/filters.py:265]: Do we need this contiguous call?

File: neutone_sdk/filters.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+        H = T.unsqueeze(-1) * torch.cat(
+            [torch.ones_like(g), -g, g, twoR * g + 1], dim=-1
+        ).reshape(seq_len, batch_size, 2, 2)
+
+        # Y = gHBx + Hs
+        gHB = g * T * torch.cat([torch.ones_like(g), g], dim=-1)
+        # [n_samples, batch_size, 2]
+        gHBx = gHB * audio.unsqueeze(-1)
+
+        Y = torch.empty(seq_len, batch_size, 2, device=audio.device)
```
Comment (bogdanteleaga) [inline neutone_sdk/filters.py:249]: Can we have this preallocated using the same mechanism as in the IIR/FIR?

File: neutone_sdk/filters.py (hunk 1 above)
Comment (hyakuchiki) [reply neutone_sdk/filters.py:265]: I put it in bc I needed it for some other projects, but it's unnecessary for this use case.

File: neutone_sdk/filters.py [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+        filt_size: Optional[int] = None,
+        filt_type: Optional[str] = None,
+    ):
+        cutoffs = self.cutoffs if cutoffs is None else cutoffs
+        sample_rate = self.sample_rate if sample_rate is None else sample_rate
+        filt_size = self.filt_size if filt_size is None else filt_size
+        filt_type = self.filt_type if filt_type is None else filt_type
+        if len(cutoffs) == 2:
+            if filt_type in ["highpass", "lowpass"]:
+                raise ValueError("only 1 cutoff value supported for this filter type")
```
Comment (bogdanteleaga) [inline neutone_sdk/filters.py:47]: ValueError(f"Only 1 cutoff value supported for filter type {filt_type}"). Same for the one below

File: neutone_sdk/filters.py [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+        cutoffs: Optional[List[float]] = None,
+        sample_rate: Optional[int] = None,
+        filt_size: Optional[int] = None,
+    ):
+        filt_type = self.filt_type if filt_type is None else filt_type
+        cutoffs = self.cutoffs if cutoffs is None else cutoffs
+        sample_rate = self.sample_rate if sample_rate is None else sample_rate
+        filt_size = self.filt_size if filt_size is None else filt_size
+        if len(cutoffs) == 2:
+            if filt_type in ["highpass", "lowpass"]:
```
Comment (bogdanteleaga) [inline neutone_sdk/filters.py:55]: FilterType.HIGHPASS, FilterType.LOWPASS, and the two below.

File: neutone_sdk/filters.py [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+        sample_rate = self.sample_rate if sample_rate is None else sample_rate
+
+        cutoff = max(min(cutoff, sample_rate / 2 - 1e-4), 1e-4)
+        resonance = max(resonance, 1e-4)
+        # frequency warping
+        self.g = torch.ones(1, 1, 1, device=self.g.device) * math.tan(
+            math.pi / sample_rate * cutoff
+        )
+        self.twoR = torch.ones(1, 1, 1, device=self.twoR.device) / resonance
+        if filt_type == "lowpass":
```
Comment (bogdanteleaga) [inline neutone_sdk/filters.py:161]: Same for these

**Review by bogdanteleaga — COMMENTED** [review]
Ok, just two more small things and let's merge it!

File: examples/example_rave_v1_prefilter.py [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+        type=str,
+        default=None,
+        help="directory of sounds to use as example input.",
+    )
+    args = parser.parse_args()
+    root_dir = Path(args.folder) / args.output
+
+    # wrap it
+    model = torch.jit.load(args.input)
+    # filter to be applied before model
```
Comment (bogdanteleaga) [inline examples/example_rave_v1_prefilter.py:169]: You forgot these comments

**Comment by bogdanteleaga** [comment]
Thank you!

**Pull request merged**